### PR TITLE
[RDY] main.c: warn about -e flag conflicting with --embed

### LINKED ISSUE #10031: Segfault with "nvim -u NONE --headless --embed -es"
```
(gdb) bt
#0  0x000055ad021a832a in linecopy (grid=0x55ad02447320 <default_grid>, to=21, from=from@entry=22, col=0, width=width@entry=80) at ../src/nvim/screen.c:6209
#1  0x000055ad021aa0a0 in grid_del_lines (grid=<optimized out>, row=21, line_count=1, end=24, col=<optimized out>, width=80) at ../src/nvim/screen.c:6353
#2  0x000055ad02224064 in msg_scroll_up () at ../src/nvim/message.c:2075
#3  0x000055ad02224263 in msg_puts_display (str=0x7ffe12247a00 "\n", maxlen=-1, attr=0, recurse=0) at ../src/nvim/message.c:1915
#4  0x000055ad02228cca in msg_puts_attr_len (str=str@entry=0x7ffe12247a00 "\n", len=len@entry=-1, attr=attr@entry=0) at ../src/nvim/message.c:1818
#5  0x000055ad02228dbd in msg_puts_attr (s=s@entry=0x7ffe12247a00 "\n", attr=attr@entry=0) at ../src/nvim/message.c:1764
#6  0x000055ad02228f72 in msg_putchar_attr (c=c@entry=10, attr=attr@entry=0) at ../src/nvim/message.c:1259
#7  0x000055ad02228fb2 in msg_putchar (c=c@entry=10) at ../src/nvim/message.c:1244
#8  0x000055ad0222959a in msg_start () at ../src/nvim/message.c:1208
#9  0x000055ad02277e1c in gotocmdline (clr=clr@entry=1) at ../src/nvim/ex_getln.c:3535
#10 0x000055ad0227c2a3 in command_line_enter (firstc=firstc@entry=58, count=count@entry=1, indent=indent@entry=0) at ../src/nvim/ex_getln.c:315
#11 0x000055ad0227c735 in getcmdline (firstc=firstc@entry=58, count=count@entry=1, indent=indent@entry=0) at ../src/nvim/ex_getln.c:2021
#12 0x000055ad0227c78a in getexline (c=58, cookie=<optimized out>, indent=0) at ../src/nvim/ex_getln.c:2205
#13 0x000055ad0228ef87 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, fgetline=0x55ad0227c766 <getexline>, cookie=cookie@entry=0x0, flags=flags@entry=0)
    at ../src/nvim/ex_docmd.c:512
#14 0x000055ad02293113 in do_exmode (improved=<optimized out>) at ../src/nvim/ex_docmd.c:229
#15 0x000055ad02214732 in normal_check (state=0x7ffe122483c0) at ../src/nvim/normal.c:1349
#16 0x000055ad021796e4 in state_enter (s=0x7ffe122483c0) at ../src/nvim/state.c:28
#17 0x000055ad022156d1 in normal_enter (cmdwin=<optimized out>, noexmode=<optimized out>) at ../src/nvim/normal.c:462
#18 0x000055ad0224153d in main (argc=<optimized out>, argv=<optimized out>) at ../src/nvim/main.c:597
```

Just noticed this when adding `-es` to a test in Neomake where I am using an embedded Neovim, which apparently hangs due to waiting for hit-Enter (from `:echom` I've added for debugging).
The test is at https://github.com/neomake/neomake/blob/6e5ff157a678812454d8bf465b630875f9a8c476/tests/compat.vader#L80-L86 - would be interested in how to ignore hit-enter there in general.

NVIM v0.4.0-744-ga541c681f

### PR DESCRIPTION
**Pull request by glacambre**
Closes https://github.com/neovim/neovim/issues/10031.

### CODE REVIEW

**Comment by glacambre** [comment]
Don't merge this - tests will fail.

**Comment by bfredl** [comment]
I don't think all of these are correct, interactive ex mode `-e` and `-E` is still usable with an `--embed` UI.

**Comment by glacambre** [comment]
You're absolutely right, it is incorrect (it makes tests fail). I don't understand why `-e --embed` works though. Shouldn't these two flags fight over stdin? One means that stdin will be text and the other that stdin will be msgpack-rpc.

**Comment by bfredl** [comment]
>  One means that stdin will be text and the other that stdin will be msgpack-rpc.

No. plain `-e` / `-E` just means start in ex mode instead of normal mode, it still uses UI code for input/output. It is equivalent to typing `gQ` just after starting.  Only `-es` / `-Es` is changing the meaning of stdin really.

File: src/nvim/main.c [hunk 1]
```diff
@@ -270,6 +270,13 @@ int main(int argc, char **argv)
 
   if (embedded_mode) {
     const char *err;
+    if (exmode_active) {
+      // exmode_active being different from 0 here means that -e/-E (read from
+      // stdin) flags are present
+      mch_errmsg(_("Error: flags -e, -E, -es and -Es are incompatible with "
+            "--embed.\n"));
+      mch_exit(2);
+    }
```
Comment (justinmk) [inline src/nvim/main.c]: `silent_mode`  signals `-es`/`-Es`.

Could be done in `command_line_scan()`, to avoid some noise in `main()`.

**Comment by glacambre** [comment]
>plain -e / -E just means start in ex mode instead of normal mode

I see, thanks for the explanation. I think the man page's description is a bit confusing, although I don't have any suggestions as to how to improve it.

I also wonder, should `--embed -` be forbidden, since it also tries to use stdin in two different ways?

**Comment by bfredl** [comment]
> I also wonder, should --embed - be forbidden, since it also tries to use stdin in two different ways?

We will address this in #10071. There seems to be a difference of behavior between `cat ... | nvim` and `cat ... | nvim -`, so we might want to forward the `-` to the `--embed` instance still.

File: src/nvim/main.c (hunk 1 above)
Comment (glacambre) [reply src/nvim/main.c]: In my opinion it's better to keep the code that parses command line arguments and the code that acts on the parsed representation separate. But I don't care enough and I'm not the one who has to maintain neovim (for now :p) so I won't argue :).

Comment (justinmk) [reply src/nvim/main.c]: maybe, but command_line_scan() has numerous such cases already--so I was thinking it's more consistent to do it in there.

File: src/nvim/main.c [hunk 2]
```diff
@@ -1217,6 +1217,11 @@ static void command_line_scan(mparm_T *parmp)
     }
   }
 
+  if (embedded_mode && exmode_active && silent_mode) {
+    mch_errmsg(_("Error: flags -es and -Es are incompatible with --embed.\n"));
+    mch_exit(2);
```
Comment (justinmk) [inline src/nvim/main.c]: "Error:" prefix is not consistent with existing startup failure messages, cf. https://github.com/neovim/neovim/blob/3afb397407af/src/nvim/main.c#L136-L142

So I removed it. Also, other messages use the term "arguments", not "flags".

Also `mainerr()` should be used.

**Comment by justinmk** [comment]
Merged. Thank you @glacambre

**Pull request closed without merging**